### PR TITLE
Adding validation to avoid null pointer exceptions

### LINKED ISSUE #15733: Possible inconsistency of job not found in DbJobState#visitJob
**Description**

Now that that we have an asynchronous scheduler in the engine, it can happen that we schedule a task to visit some job state, but the job is completed or cancelled in the meantime. This leads to us visiting a non-existent job. That's safe to do, but we log the following at **error** level:

```
Expected to find job with key 221234128273, but no job found
```

It's unclear to me yet if this is a bug or if this is purely noise, primarily because the method `visitJob` is called only via the iterative methods:

- `DbJobStat#forEachTimedOutEntry`
- `DbJobState#forEachActivatableJob`
- `DbJobState#findBackedOffJobs`

Each of these methods should be called in a transaction, such that `visitJob` would also be in the same transaction, so if they did not see a job, it means the following:

- We have some broken transaction boundaries (e.g. somehow we're using two transactions here, or the other update is done in multiple, non-atomic transactions)
- We have garbage in the activatable/timed out/backed off columns for jobs which don't exist anymore.
- I'm missing something and this is fine, and we're just unnecessarily logging at the error level.

The first one is probably a high severity, or even critical bug, but the second one is more like a mid.

### LINKED ISSUE #16609: NPE in PendingProcessMessageSubscriptionChecker
**Describe the bug**

```
"Uncaught exception in 'AsyncProcessingScheduleActor-3' in phase 'STARTED'. Continuing with next job."

java.lang.NullPointerException: Cannot invoke "io.camunda.zeebe.engine.state.message.ProcessMessageSubscription.isOpening()" because "subscription" is null
	at io.camunda.zeebe.engine.processing.message.PendingProcessMessageSubscriptionChecker.sendPendingCommand(PendingProcessMessageSubscriptionChecker.java:86) ~[zeebe-workflow-engine-8.5.0-SNAPSHOT.jar:8.5.0-SNAPSHOT]
	at io.camunda.zeebe.engine.state.message.DbProcessMessageSubscriptionState.visitPending(DbProcessMessageSubscriptionState.java:175) ~[zeebe-workflow-engine-8.5.0-SNAPSHOT.jar:8.5.0-SNAPSHOT]
	at io.camunda.zeebe.engine.processing.message.PendingProcessMessageSubscriptionChecker.checkPendingSubscriptions(PendingProcessMessageSubscriptionChecker.java:79) ~[zeebe-workflow-engine-8.5.0-SNAPSHOT.jar:8.5.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:96) ~[zeebe-scheduler-8.5.0-SNAPSHOT.jar:8.5.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:47) [zeebe-scheduler-8.5.0-SNAPSHOT.jar:8.5.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:122) [zeebe-scheduler-8.5.0-SNAPSHOT.jar:8.5.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:130) [zeebe-scheduler-8.5.0-SNAPSHOT.jar:8.5.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:108) [zeebe-scheduler-8.5.0-SNAPSHOT.jar:8.5.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:227) [zeebe-scheduler-8.5.0-SNAPSHOT.jar:8.5.0-SNAPSHOT]
"
```

**Impact**
Only happened once. The partition was healthy afterwards, and continued to process with no issues.
Not sure if `PendingProcessMessageSubscriptionChecker` continued to run. 

**Logs**

[link](https://console.cloud.google.com/logs/query;cursorTimestamp=2024-02-26T13:04:16.615906160Z;endTime=2024-02-26T13:34:46.615Z;pinnedLogId=2024-02-26T13:04:16.615906160Z%2Fttlan475z65chldw;query=logName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.location%3D%22europe-west1%22%0Aresource.labels.namespace_name%3D%22e854f9ed-af8b-4cc2-a94f-50b9c14de75e-zeebe%22%0Aresource.labels.cluster_name%3D%22worker-chaos-1%22%0Aresource.labels.container_name%3D%22zeebe%22%0Aresource.labels.project_id%3D%22camunda-saas-int-chaos%22%0Aresource.labels.pod_name%3D%22zeebe-2%22%0AjsonPayload.context.partitionId%3D%223%22%0AjsonPayload.context.%22actor-scheduler%22%3D%22Broker-2%22%0A-jsonPayload.context.loggerName%3D%22io.atomix.raft.roles.LeaderAppender%22%0Atimestamp%3D%222024-02-26T13:04:16.615906160Z%22%0AinsertId%3D%22ttlan475z65chldw%22;startTime=2024-02-26T12:34:46.615Z?project=camunda-saas-int-chaos) 

There were no relevant logs related to this around that time. So I didn't collect them. 

**Environment:**
- Zeebe Version: 8.5.0-SNAPSHOT-main-e5c82370

### PR DESCRIPTION
**Pull request by ana-vinogradova-camunda**
## Description

Adding a Logger to `DbProcessMessageSubscriptionState` in order to log warning in cases where the subscription is not found. Please see [this bug](https://github.com/camunda/zeebe/issues/16609)

Changing ERROR log message to WARN in `DbJobState` to align with other checkers.

## Related issues

closes #16609
closes #15733 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java [hunk 1]
```diff
@@ -430,7 +430,7 @@ public long findBackedOffJobs(final long timestamp, final BiPredicate<Long, JobR
   boolean visitJob(final long jobKey, final BiPredicate<Long, JobRecord> callback) {
     final JobRecord job = getJob(jobKey);
     if (job == null) {
-      LOG.error("Expected to find job with key {}, but no job found", jobKey);
+      LOG.warn("Expected to find job with key {}, but no job found", jobKey);
```
Comment (megglos) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java:433]: 💭  would also be worth a separate commit as both changes are related by also independent

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/message/DbProcessMessageSubscriptionState.java [hunk 2]
```diff
@@ -171,7 +174,12 @@ public void visitPending(final long deadline, final ProcessMessageSubscriptionVi
               pendingSubscription.elementInstanceKey(),
               BufferUtil.wrapString(pendingSubscription.messageName()),
               pendingSubscription.tenantId());
-
+      if (subscription == null) {
+        LOG.warn(
+            "Expected to find subscription with key {}, but no subscription found",
+            pendingSubscription.elementInstanceKey());
```
Comment (megglos) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/message/DbProcessMessageSubscriptionState.java:180]: ❌ actually all parameters are used for lookup, let's include them in the log to be able to reconstruct the whole key
```suggestion
            "Expected to find subscription with key {} messageName {} tenantId: {}, but no subscription found",
            pendingSubscription.elementInstanceKey(), messageName, tenantId);
```

**Review by megglos — APPROVED** [review]
Nice! One suggestion to extend the log statement on the missing subscription, pre-approving.

**Comment by ana-vinogradova-camunda** [comment]
Closing PR in favour of [this PR](https://github.com/camunda/zeebe/pull/16873).
